### ISSUE TITLE
Support pagination

### ISSUE THREAD

**Issue by mwildehahn** [mwildehahn/hack-graphql#38]
https://graphql.org/learn/pagination/

**Comment by mwildehahn** [comment]
Specifically: https://relay.dev/graphql/connections.htm

**Comment by ianhoffman** [comment]
@mwildehahn do you want to take this on? I was thinking about doing it but could also keep just churning out validation rules 🤷‍♂️ .

This would be suuuuper cool to have for the MVP.

**Comment by mwildehahn** [comment]
Do it! Agree this would be really cool. I think it's worth discussing what this looks like, ie is this something like:

```
<<GraphQL\Field('friends', 'list of friends')>>
public function getFriends(GraphQL\PaginationArgs $args): GraphQL\Connection<User> {
}
```

**Comment by ianhoffman** [comment]
That's pretty close to what I was envisioning!

**Comment by mwildehahn** [comment]
I imagine we'd need something like these helpers too: https://github.com/graphql/graphql-relay-js/blob/main/src/connection/arrayConnection.js#L19

**Comment by ianhoffman** [comment]
> I think it's worth discussing what this looks like

Me too. I'm going to draw up a proposal for the API and run it by you and @jjergus, hopefully today or tomorrow.

**Comment by ianhoffman** [comment]
The TL;DR is gonna be that you can subclass `Connection<T>` and return your subclass from a GraphQL field to get automatic pagination.

**Comment by mwildehahn** [comment]
That will be *amazing*

**Comment by jjergus** [comment]
If you have any framework that you currently use internally for querying data with slicing/sorting/filtering/etc. then probably the most important goal here is to make sure this works well with it.

If your framework already has some kind of object that represents a "pending query" (i.e. an object that contains no actual data, just information about what needs to be fetched from a DB, like cursors, filters etc.), you could use that object instead of a new `Connection<T>` type (alternatively, consider whether that object can be updated to extend `Connection<T>`).

For example, let's assume `UserQuery` represents a query to fetch some users. The GraphQL framework could recognize something like

```Hack
<<GraphQL\Field('friends')>>
public function queryFriends(...): UserQuery {
  ...
}
```

and generate a connection out of it.

**Comment by ianhoffman** [comment]
Internally we have a very simple pagination model: basically, all you need is a cursor and an offset and we can do pagination. So I think that fits quite nicely with the relay model already. 

I do think we should have a `Connection<T>` which is separate from our internal representation (which is function-, not called-, based anyway).

**Comment by jjergus** [comment]
Makes sense, if you don't already have a class that would fit naturally, a new `Connection<T>` sounds good.

**Comment by mwildehahn** [comment]
We use very little object oriented programming, but we’re starting to use
more.

We also based our cursor pagination off of relay originally (the concept of
an opaque base64 encoded string) and it will be nice to actually use it
within graphql!

**Comment by ianhoffman** [comment]
@jjergus @mh Here's a tech spec: https://paper.dropbox.com/doc/Tech-Spec-GraphQL-Connections--BLIP7PNUHGqviY95S8LmDIPdAg-tfBrXPdOj9B5gvvuoxVAN

**Comment by jjergus** [comment]
There's also an official formal specification, but I don't think it says anything you don't already know :)

https://relay.dev/graphql/connections.htm

**Comment by ianhoffman** [comment]
Yup, that (and the Relay spec) are what I went off for this.

**Comment by jjergus** [comment]
Is your plan to have one GraphQL connection type per target type (e.g. `UserConnection` for every field that returns a paginable list of `User`s) or one GraphQL connection type per field (e.g. `FriendsConnection` and `RelativesConnection` for 2 fields that both return a paginable list of `User`s)?

The latter might be useful (necessary?) if you want any "edge data", e.g. you might want a "relationship (uncle/aunt/...)" on `RelativesConnection`'s edges but not on `FriendsConnection`.

**Comment by ianhoffman** [comment]
I guess my plan (to the extent that I have one) is to leave this up to the implementor: you can return a `UserConnection::friendsOf($user->getId())`, which constructs a connection that returns friends for a user but might also be useful for other things, or you could define a `FriendsConnection`. The current proposal would easily support both approaches.

**Comment by ianhoffman** [comment]
Closed via #91
